### ISSUE TITLE
Admits object literals with incompatible types to map when using computed properties

### ISSUE THREAD

**Issue by ned** [facebook/flow#1971]
When using a map type, such as `{[key: string]: number}`, with an object literal that uses computed properties, Flow seems to be failing to detect that the types of the key or value of the object are not compatible with the map type.

``` js
// @flow

type Map = {[key: string]: number}

/* Neither the key or value types are correct, but Flow
 * still admits them.
 */
const a: Map = {[0]: "hello"};

/* Same problem.
 *
 * This is a common scenario in Redux projects, where
 * the spread operator is used to make a copy.
 *     return {...old, [id]: value}
 */
function f(): Map {
    return {[0]: "hello"};
}

/* This version DOES throw an error, for both
 * the key and value, as desired
 */
const b: Map = {};
b[0] = "hello";

/* This also correctly throws an error for the value */
const c: Map = {["0"]: "hello"};

/* As does this */
const d: Map = {"0": "hello"};

/* But this fails to throw an error */
const e: Map = {["" + "0"]: "hello"};
```

**Comment by echojc** [comment]
I've encountered a similar issue where an explicit type annotation on the value used as the key of a computed property also stops Flow from reporting incompatible values in the map.

```es6
/* @flow */

const key = 'key1';
const annotatedKey: string = 'key2';

const map: { [key: string]: string } = {
  [key]: 1,
  [annotatedKey]: 2
};

map[key] = 3;
map[annotatedKey] = 4;
```

Flow only reports errors for cases 1, 3 and 4.

```
7:   [key]: 1,
            ^ number. This type is incompatible with
6: const map: { [key: string]: string } = {
                               ^ string
11: map[key] = 3;
               ^ number. This type is incompatible with
6: const map: { [key: string]: string } = {
                               ^ string
12: map[annotatedKey] = 4;
                        ^ number. This type is incompatible with
6: const map: { [key: string]: string } = {
                               ^ string
```

**Comment by gkz** [comment]
Fixed for number keys in cfe00503857c1e3dc2566f8d56c3cf522c017b36 (all but example `e` in initial example).
